06_gpu_and_ml/yolo/yolo.py

```diff
@@ -0,0 +1,288 @@
+# ---
+# runtimes: ["runc", "gvisor"]
+# ---
+#
+# # Training and Inferencing YOLOv10
+#
+# Example by [@AnirudhRahul](https://github.com/AnirudhRahul)
+#
+# [YOLOv10](https://docs.ultralytics.com/models/yolov10/) is the most modern version of the popular "You Only Look Once" ("YOLO") computer vision model, 
+# which has been trained to perform bounding box object detection.
+#
+# In this example, we use the V10 model via the [Ultralytics package](https://docs.ultralytics.com/). We will:
+# - Fine tune on a custom dataset, in our case the DoclayNet dataset for object detection within documents.
+# - Deploy an inference endpoint.
+
+# For commercial use, sure to consult the [ultralytics software license](https://docs.ultralytics.com/#yolo-licenses-how-is-ultralytics-yolo-licensed).
+
+# ## Building Our Image
+#
+# The example uses the Ultralytics package to run training inference and the Huggingface dataset package to load DoclayNet data.
+# 
+# All packages are installed into a Pytorch 2.3.1 base image
+# using the `pip_install` function.
+#
+# In addition, we download the base YOLO model weights into the container at build-time
+# using the `run_function` function
+
+import modal
+import urllib.request, os, yaml, json, shutil, zipfile
+from pathlib import Path
+
+def load_yolo_weights():
+    from ultralytics import YOLO
+    # Choose the large size checkpoint
+    # Other model sizes can be found at: https://docs.ultralytics.com/models/yolov10/#performance
+    model = YOLO('yolov10l.pt') # This downloads the model's base weights if not already present in the image.
+    return model
+
+yolo_image = (
+    modal.Image.from_registry("pytorch/pytorch:2.3.1-cuda11.8-cudnn8-runtime")
+    .apt_install(["libgl1-mesa-glx", "libglib2.0-0"])
+    .pip_install(["ultralytics", "datasets", "opencv-python"])
```

Comment:
We like pinning to avoid drift-induced breakage.

Reply:
@erik-dunteman see the PR template for a detailed policy.

`apt_install`ed libs are generally not pinned, because they are more dependable than Python libraries.